### DISCUSSION TITLE
Why does pytest not execute the entire loop

### DISCUSSION THREAD

**Discussion by tejendrapsingh** [pytest-dev/pytest#12926] [Q&A]

**Comment by graingert** [comment]
pytest with xdist doesn't print anything out when using `-s` https://github.com/pytest-dev/pytest-xdist/issues/354

Probably this issue is in where you catch all exceptions and suppress them https://github.com/tejendrapsingh/for-help/blob/main/using-pytest-for-parallelization.py#L900

Instead you should only catch the exceptions that you expect 

If you want to catch multiple exceptions and continue a for loop you should use an exception group:

```python
exceptions = []
for x in y:
    try:
        ...
    except Exception as e:
        e.add_note("failed processing {x}")
        exceptions.append(e)
if exceptions:
    raise ExceptionGroup("multiple errors processing", exceptions)
```

These bare except statements are also bad, and incompatible with a signal handler and BaseException timeout approach: https://github.com/tejendrapsingh/for-help/blob/main/using-pytest-for-parallelization.py#L92

**Reply by graingert** [reply]
Also this is not the correct way to add metadata to an exception:

```python
except Exception as e:
        print("*** print_exception:")
        pytest.fail(f"Test failed for {data} {year_val} with error: {str(e)}")
```

You should instead do:

```python
except Exception as e:
    e.add_note(f"Test failed for {data} {year_val}")
    raise
```

**Reply by tejendrapsingh** [reply]
@graingert Thanks for these suggestions. I will edit the code with what you suggested and see if it makes a difference. Thank you very much for taking the time to go through the code. I am not a native programmer. So, always trying to find the easiest way to make things work. Pardon my ``dirty'' code writing.

**Comment by The-Compiler** [comment]
I'd suggest you don't use pytest for this at all, as you don't actually seem to be testing anything. If you use Selenium directly from Python, and then use something like [the `concurrent.futures` module](https://stackoverflow.com/a/69247787/2085149) to parallelize that, you'll have much more control over what's happening.

I also agree that your approach to exception handling likely hides bugs in your code - see e.g. [The Most Diabolical Python Antipattern – Real Python](https://realpython.com/the-most-diabolical-python-antipattern/).
